[FIR]: fix KT-42351

juan-chen opened the pull request:
FirVisibilityChecker::isVisible checked if a private declaration can be
accessed at a use site by matching class ids of the private
declaration's owner with the use site's containing class
declarations. When the private declaration is defined in an enum
entry and used in the same entry, its owner class id has the enum
entry name, but the use site is in an FirAnonymousObject, which has
"anonymous" as the class id. This causes visibility check to fail.

This PR adds handling of enum entries in isVisible: it checks if any
containing FirAnonymousObject of the use site contains the private
declaration.

demiurg906 commented on compiler/fir/resolve/src/org/jetbrains/kotlin/fir/FirVisibilityChecker.kt:
```diff
@@ -67,6 +67,8 @@ abstract class FirVisibilityChecker : FirSessionComponent {
                         // Member: visible inside parent class, including all its member classes
                         canSeePrivateMemberOf(containingDeclarations, ownerId, session)
+                                // Handle enum entries, as they are represented as anonymous objects (KT-42351)
+                                || containingDeclarations.any { it is FirAnonymousObject && it.declarations.contains(declaration) }
```
This check looks like ad-hock hack

demiurg906 commented:
Hello @juan-chen. This issue happens not because of `FirVisiblityChecker` don't support members of enum entries, but because of raw fir builder incorrectly builds callable ids for them.

Check that such case is working:
```kotlin
fun test() {
    val x = object : Any {
        private fun foo() {}
        
        fun bar() {
            foo()
        }
    }
}
```

I can't accept your fix, but you can dig this problem deeper

demiurg906 commented:
We discussed this problem in a team and decided that this problems should be fixed in RawFirBuilder next way: self type which passed to declarations of enum entry should have special class id same as for regular anonymous objects

demiurg906 commented:
@juan-chen this fix looks much cleaner, thanks.

Can you please also support it in `lightTree2fir` builder?
Also check failing FIR tests in your branch:
- `Fir2IrTextTestGenerated$Declarations.testInlineCollectionOfInlineClass`
- `TreesCompareTest.testCompareDiagnostics`

juan-chen commented:
Thanks Dmitriy!

> Can you please also support it in `lightTree2fir` builder?

Done.

> Also check failing FIR tests in your branch:
> 
> * `Fir2IrTextTestGenerated$Declarations.testInlineCollectionOfInlineClass`

This failure seems unrelated to the PR, fixed by beac3757 already.

> * `TreesCompareTest.testCompareDiagnostics`

Fixed

demiurg906 commented:
Merged manually
